feat(docdb): high level constrcuts for db clusters and instances (#6511)

jusiskin opened the pull request:
## Commit Message
feat(docdb): high level constrcuts for db clusters and instances (#6511)

- CDK users who want to want to use Amazon DocumentDB previously had to use the low-level auto-generated constructs (CFN Resources)
- Adds higher-level constructs for
  - `AWS::DocDB::DBCluster`
  - `AWS::DocDB::DBInstance`
  - `AWS::DocDB::DBClusterParameterGroup`

fixes #6511

## End Commit Message

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

<!-- 
Please read the contribution guidelines and follow the pull-request checklist:
https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md
 -->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: c8a567cee86cfc31030501b1e328b57d73535316
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e5f1a119-5445-45b2-ae58-6b10bfd4793c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 9b7e3dca37310f553fa2f207d861a1a1f0bed42e
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d2e85cc1-8d84-4f2f-97e3-37b36a389cd5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jusiskin commented:
Looks like there is a conflict with `yarn.lock`. Just in case the reviewer (@iliapolo) is in mid-review, I will wait until I get the go ahead to rebase the pull-request.

jusiskin commented:
Just wanting to check-in on this PR. Is there anything I can do to help move it forward?

iliapolo commented:
Hi @jusiskin - Thanks for this PR! Looks like an amazing addition. I will start reviewing it soon.

In the meantime, you can resolve any conflicts 👍

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 6a2b8544026c188af2aab2d15b853cea813f5d26
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d2acd05c-c8d2-4176-bcd7-00e3cdc83a29%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

iliapolo commented on packages/@aws-cdk/aws-docdb/lib/endpoint.ts:
```diff
@@ -0,0 +1,81 @@
+   *
+   * @returns {string} An (un)resolved string representation of the endpoint's port number
+   */
+  public portAsString(): string {
```
Im not a hundred percent convinced here. Can you share an example usage of string embedding that doesn't work without this?

jusiskin replied:
Consider using the port as a string in an instance's user data script.

```ts
import * as cdk from '@aws-cdk/core';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as docdb from '@aws-cdk/aws-docdb';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'MyStack');
const vpc = new ec2.Vpc(stack, 'Vpc');

const cluster = new docdb.DatabaseCluster(stack, 'Database', {
  masterUser: {
    username: 'docdb'
  },
  instanceProps: {
    instanceType: ec2.InstanceType.of(ec2.InstanceClass.R5, ec2.InstanceSize.LARGE),
    vpc
  },
  removalPolicy: cdk.RemovalPolicy.DESTROY
});

// Create instance to run my application
const instance = new ec2.Instance(stack, 'Instance', {
  vpc,
  instanceType: ec2.InstanceType.of(ec2.InstanceClass.T3, ec2.InstanceSize.MICRO),
  machineImage: new ec2.AmazonLinuxImage()
});

// Allow network traffic from the instance
cluster.connections.allowDefaultPortFrom(instance);

// Start my app on launch and connect to the DocumentDB cluster
instance.userData.addCommands(
  'pip install mysuperawesomeapp',
  `mysuperawesomeapp --dbhost ${cluster.clusterEndpoint.hostname} --dbport ${cluster.clusterEndpoint.portAsString()}`
);
```

CDK does not resolve numeric tokens within strings. If you use `cluster.clsuterEndpoint.port` in the userdata command string it gets left as an unresolved token.

jogold commented on packages/@aws-cdk/aws-docdb/lib/cluster.ts:
```diff
@@ -0,0 +1,454 @@
+    if (props.instanceProps.securityGroup) {
+      securityGroup = props.instanceProps.securityGroup;
+    } else {
+      securityGroup = new ec2.SecurityGroup(this, 'SecurityGroup', {
```
The construct implements `IConnectable`. The idea is to allow to use the `connections` API https://github.com/aws/aws-cdk/tree/master/packages/%40aws-cdk/aws-ec2#allowing-connections.

jusiskin replied:
Yes - what @jogold said. Otherwise using `cluster.connections` will not work without supplying a security group in the props.

jusiskin commented on packages/@aws-cdk/aws-docdb/lib/database-secret.ts:
```diff
@@ -0,0 +1,62 @@
+      encryptionKey: props.encryptionKey,
+      // Source: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-docdb-dbcluster.html
+      // CloudFormation currently limits the DocDB master password to 41 characters with:
+      // Length of value for property {/MasterUserPassword} is greater than maximum allowed length {41}
```
At the time I originally wrote the code, I started with what the CloudFormation documentation and the DocumentDB documentation specified and set the password length to 100 characters. When I tested deploying this, I got the error message in the comment above for the `AWS::DocDB::DBCluster` resource. The maximum password length appears to not match what is documented.

jusiskin commented on packages/@aws-cdk/aws-docdb/lib/database-secret.ts:
```diff
@@ -0,0 +1,62 @@
+   *
+   * @default Secretsmanager will generate a physical name for the secret
+   */
+  readonly secretName?: string;
```
Beyond the master user, you can attach additional users (and corresponding secrets) to the cluster. There's a handy multi user rotation example in the included `README.md` file.

jusiskin commented on packages/@aws-cdk/aws-docdb/lib/cluster.ts:
```diff
@@ -0,0 +1,454 @@
+      securityGroups: [securityGroup]
+    });
+
+    this.node.defaultChild = cluster;
```
Looks like this can be removed. I must have borrowed this from another package in the CDK repo. I traced this and found that it is not necessary because the cluster already uses the `"Resource"` construct ID.

See https://github.com/aws/aws-cdk/blob/cbe735e4c5a740c7e6f304bcd9aa867e4b44a9d8/packages/%40aws-cdk/core/lib/construct-compat.ts#L323-L343

jusiskin commented on packages/@aws-cdk/aws-docdb/lib/cluster.ts:
```diff
@@ -0,0 +1,454 @@
+   *
+   * @default - A name is automatically generated.
+   */
+  readonly dbClusterName?: string;
```
This is named to satisfy the `awslint:props-physical-name` rule. See https://github.com/aws/aws-cdk/blob/cbe735e4c5a740c7e6f304bcd9aa867e4b44a9d8/tools/awslint/lib/rules/resource.ts#L221-L229

jusiskin commented on packages/@aws-cdk/aws-docdb/lib/cluster.ts:
```diff
@@ -0,0 +1,454 @@
+      preferredMaintenanceWindow: props.preferredMaintenanceWindow,
+      // Encryption
+      kmsKeyId: props.kmsKey && props.kmsKey.keyArn,
+      storageEncrypted: props.kmsKey ? true : props.storageEncrypted,
```
If this is the case, then perhaps I should just remove the `storageEncrypted` prop and leave out the validation. Maybe I'm missing something, but the encryption seems to require a KMS key and encryption can be implied when one is supplied. On that basis, how would you feel about changing the cluster props to:

```ts
interface ClusterEncryptionProps {
  readonly kmsKey: kms.IKey;
}

interface DatabaseClusterProps {
  /**
   * Optional encryption settings for the cluster.
   *
   * @default The database cluster is not encrypted
   */
  readonly encryption?: ClusterEncryptionProps;
  // ...
}
```

I understand this is creating an interface for a single attribute, but it makes the calling code a little more readable:

```ts
const cluster = DatabaseCluster(stack, 'Cluster', {
  // ...
  encryption: {
    kmsKey
  }
});
```

jusiskin commented on packages/@aws-cdk/aws-docdb/lib/cluster.ts:
```diff
@@ -0,0 +1,454 @@
+    });
+
+    this.clusterIdentifier = cluster.ref;
+    this.clusterResourceIdentifier = cluster.attrClusterResourceId;
```
From the [DocDB developer guide](https://docs.aws.amazon.com/documentdb/latest/developerguide/API_DBCluster.html):

>   DBClusterIdentifier
>   Contains a user-supplied cluster identifier. This identifier is the unique key that identifies a cluster.

Also:

>   DbClusterResourceId
>   The AWS Region-unique, immutable identifier for the cluster. This identifier is found in AWS CloudTrail log entries whenever the AWS KMS key for the cluster is accessed.

From the [CloudFormation reference for `AWS::DocDB::DBCluster`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-docdb-dbcluster.html#aws-resource-docdb-dbcluster-return-values):

>   ClusterResourceId
>   The resource id for the cluster; for example: cluster-ABCD1234EFGH5678IJKL90MNOP. The cluster ID uniquely identifies the cluster and is used in things like IAM authentication policies.

It appears that there are two separate identifiers. `DBClusterIdentifier` is a unique user-supplied name - presumably to present a human-readable identify to AWS users. `DbClusterResourceId` seems to be the backend ID for things like CloudTrail and IAM.

I'd imagine there would be use-cases for both attributes in CDK apps.

jusiskin commented on packages/@aws-cdk/aws-docdb/lib/cluster.ts:
```diff
@@ -0,0 +1,454 @@
+    }
+
+    // Create the instances
+    const instanceCount = props.instances != null ? props.instances : DatabaseCluster.DEFAULT_NUM_INSTANCES;
```
I guess this is not logically equivalent when `instanceCount == 0`. Should we apply the default if the user supplies 0, or fail with an exception as done in the line below?

jusiskin commented on packages/@aws-cdk/aws-docdb/lib/cluster.ts:
```diff
@@ -0,0 +1,454 @@
+        defaultPort: this.defaultPort
+      });
+      public readonly clusterIdentifier = attrs.clusterIdentifier;
+      public readonly instanceIdentifiers: string[] = [];
```
Good catch! I will correct this

jusiskin commented on packages/@aws-cdk/aws-docdb/lib/cluster.ts:
```diff
@@ -0,0 +1,454 @@
+      dbSubnetGroupName: subnetGroup.ref,
+      port: props.port,
+      vpcSecurityGroupIds: [this.securityGroupId],
+      dbClusterParameterGroupName: props.parameterGroup && props.parameterGroup.parameterGroupName,
```
Fixed

jusiskin commented on packages/@aws-cdk/aws-docdb/lib/cluster.ts:
```diff
@@ -0,0 +1,454 @@
+  readonly parameterGroup?: IClusterParameterGroup;
+
+  /**
+   * The removal policy to apply when the cluster and its instances are removed
```
Good suggestion. I've applied it

jusiskin commented on packages/@aws-cdk/aws-docdb/lib/cluster.ts:
```diff
@@ -0,0 +1,454 @@
+    const { subnetIds, internetConnectivityEstablished } = this.vpc.selectSubnets(this.vpcSubnets);
+
+    // Cannot test whether the subnets are in different AZs, but at least we can test the amount.
+    if (subnetIds.length < 2) {
```
Done

jusiskin commented:
Hi @iliapolo, thanks for your review. I've rebased, applied all of the feedback, squashed the commits and pushed it back up. This should be ready for another review pass.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: bbfe1c2dce4c12aa645804dc9db65f2fcafff6b7
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bddcc700-fb7c-426a-bbab-0ea35a62b3bb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

iliapolo commented on packages/@aws-cdk/aws-docdb/lib/parameter-group.ts:
```diff
@@ -0,0 +1,86 @@
+import { Construct, IResource, Resource } from '@aws-cdk/core';
```
I think we can model the parameter group in a way that is clearer to users. From my own experience, chasing down these parameters can be annoying, and also somewhat error prone. 

Looking at the CloudFormation [docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-docdb-dbclusterparametergroup.html#aws-resource-docdb-dbclusterparametergroup--examples), i see three parameters that can be used (there might be more - but just as an example).

```yaml
Parameters: 
  audit_logs: "disabled"
  tls: "enabled"
  ttl_monitor: "enabled"
```   

What if we provided an api like so, on the `DatabaseCluster` construct:

```typescript
cluster.enableAuditLogs()
cluster.enableTls()
cluster.enableTTLMonitor()
```

Internally, these methods will collect the parameters and create and wire the `CfnParameterGroup` on the user's behalf.

We can also have a generic method like:

```typescript
cluster.addParameter(key, value)
```
To support any parameters that we don't have explicit API's for. This will also sets a good example for subsequent contributions, until we eventually cover all these parameters with proper API's, which is heaven. 

It would also be amazing if we could somehow map the `engineVersion` to the `family` of the parameter group, without the user needing to even think about it. 

I know it's not a trivial change, but I think it's not that big also, and provides a great experience. What do you think?

jusiskin replied:
I like where you're going with this suggestion, but it does expand the scope of this pull request which focuses on adding support for clusters and instances. Parameter groups are a useful feature of DocumentDB, but are not strictly required to use the service.

As-is, the current API allows users of the `ParameterGroup` L2 construct to supply arbitrary parameters. While the set of supported parameters are not easily discoverable and requires work on their part to find the proper documentation, it is not a show-stopper to being able to use any of the features.

I agree on all your arguments about setting an example and providing a great experience, but would you be opposed to leaving this work to a future PR? I think CDK users would benefit from L2 DocumentDB constructs in their current state and could stand to wait a bit longer for the improved APIs. In the mean-time, I can provide a handy link to the parameter reference documentation in the TSDoc string.

iliapolo replied:
@jusiskin I hear you, Lets move forward with the PR as is. I will follow it up with an issue regarding the `ParameterGroup` 👍

jusiskin replied:
Thanks. I'm just integrating the rest of your feedback as we speak. Hopefully we can merge this today :)

jusiskin commented on packages/@aws-cdk/aws-docdb/test/cluster.test.ts:
```diff
@@ -0,0 +1,706 @@
+    // GIVEN
+    const stack = testStack();
+
+    const cluster = DatabaseCluster.fromDatabaseClusterAttributes(stack, 'Database', {
```
done

jusiskin commented on packages/@aws-cdk/aws-docdb/test/cluster.test.ts:
```diff
@@ -0,0 +1,706 @@
+  });
+
+});
+
```
done

jusiskin commented on packages/@aws-cdk/aws-docdb/lib/cluster.ts:
```diff
@@ -0,0 +1,453 @@
+  readonly instanceProps: InstanceProps;
+
+  /**
+   * A daily time range in 24-hours UTC format in which backups preferably execute.
```
done

jusiskin commented on packages/@aws-cdk/aws-docdb/package.json:
```diff
@@ -83,22 +83,43 @@
   "dependencies": {
+    "@aws-cdk/aws-efs": "0.0.0",
+    "@aws-cdk/aws-ec2": "0.0.0",
+    "@aws-cdk/aws-events": "0.0.0",
```
removed

jusiskin commented on packages/@aws-cdk/aws-docdb/package.json:
```diff
@@ -83,22 +83,43 @@
+    "@aws-cdk/aws-ec2": "0.0.0",
+    "@aws-cdk/aws-events": "0.0.0",
+    "@aws-cdk/aws-kms": "0.0.0",
+    "@aws-cdk/aws-lambda": "0.0.0",
```
removed

jusiskin commented on packages/@aws-cdk/aws-docdb/package.json:
```diff
@@ -83,22 +83,43 @@
+    "@aws-cdk/aws-kms": "0.0.0",
+    "@aws-cdk/aws-lambda": "0.0.0",
+    "@aws-cdk/aws-secretsmanager": "0.0.0",
+    "@aws-cdk/aws-sqs": "0.0.0",
```
removed

jusiskin commented on packages/@aws-cdk/aws-docdb/package.json:
```diff
@@ -68,8 +68,8 @@
       "global": {
-        "branches": 60,
-        "statements": 80
+        "branches": 95,
```
The tooling didn't seem to like having the `"jest"` top-level key at all in `package.json` and just wants:

```js
{
    // ...
    "cdk-build": {
        "jest": true
    }
    // ...
}
```

All the tests pass, so it looks to be good.

jusiskin commented on packages/@aws-cdk/aws-docdb/package.json:
```diff
@@ -83,22 +83,43 @@
     "node": ">= 10.12.0"
   },
+  "awslint": {
+    "exclude": [
```
done

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 71fd6f4cb9a55586ccaa5b9fdec4660158869e34
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d410e25e-26d8-49a1-84a4-7bc5cf00c35e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 7cc5603285ad46623f2ad9949f32239b526be404
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=29da50e2-fa03-49e0-b242-12038a016ee6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: d4ebb38ce5c05ac51505e0974c2e1b9b3a112531
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=036e9f47-b9db-466e-9e99-7fbff240ef1b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 02315545dd3692b89af0789a8e294a0b351bb4bb
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=eaf1b97f-3b7c-41a0-95bd-d4abf184d18e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 9322b41b3d72a65b5ce59a386317460f6cb9246f
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=89ea1f2a-9208-4132-b211-7f0aa9d075cb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
